src/Functions/splitByRegexp.cpp

```diff
@@ -148,11 +150,59 @@ class SplitByRegexpImpl
 
 using FunctionSplitByRegexp = FunctionTokens<SplitByRegexpImpl>;
 
+/// Fallback splitByRegexp to splitByChar when its 1st argument is a trivial char for better performance
+class SplitByRegexpOverloadResolver : public IFunctionOverloadResolver
+{
+public:
+    static constexpr auto name = "splitByRegexp";
+    static FunctionOverloadResolverPtr create(ContextPtr context) { return std::make_unique<SplitByRegexpOverloadResolver>(context); }
+
+    explicit SplitByRegexpOverloadResolver(ContextPtr context_) : context(context_), split_by_regexp(FunctionSplitByRegexp::create(context))
```

Comment:
I guess we should initialize `split_by_regexp` only on-demand, i.e. if we don't go for the fallback. Initialization of `split_by_regexp` accesses settings which needs a lock, so it is not for free.

Reply:
Notice that `split_by_regexp` is needed in `SplitByRegexpOverloadResolver::getReturnTypeImpl`. There is no chance that `split_by_regexp` is initialized lazily.